// homebridge-ring/lib/homebridge/co-alarm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoAlarm = void 0;
const base_device_accessory_1 = require("./base-device-accessory");
const hap_1 = require("./hap");
class CoAlarm extends base_device_accessory_1.BaseDeviceAccessory {
    constructor(device, accessory, logger, config) {
        super();
        this.device = device;
        this.accessory = accessory;
        this.logger = logger;
        this.config = config;
        const { Characteristic: { CarbonMonoxideDetected }, Service: { CarbonMonoxideSensor }, } = hap_1.hap;
        this.registerCharacteristic({
            characteristicType: CarbonMonoxideDetected,
            serviceType: CarbonMonoxideSensor,
            getValue: (data) => {
                return data.alarmStatus === 'active'
                    ? CarbonMonoxideDetected.CO_LEVELS_ABNORMAL
                    : CarbonMonoxideDetected.CO_LEVELS_NORMAL;
            },
        });
        this.initSensorService(CarbonMonoxideSensor);
    }
}
exports.CoAlarm = CoAlarm;
