// homebridge-ring/lib/homebridge/smoke-co-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmokeCoListener = void 0;
const base_device_accessory_1 = require("./base-device-accessory");
const hap_1 = require("./hap");
class SmokeCoListener extends base_device_accessory_1.BaseDeviceAccessory {
    constructor(device, accessory, logger, config) {
        super();
        this.device = device;
        this.accessory = accessory;
        this.logger = logger;
        this.config = config;
        const { Characteristic: { SmokeDetected, CarbonMonoxideDetected }, Service: { SmokeSensor, CarbonMonoxideSensor }, } = hap_1.hap;
        this.registerCharacteristic({
            characteristicType: SmokeDetected,
            serviceType: SmokeSensor,
            getValue: (data) => {
                return data.smoke && data.smoke.alarmStatus === 'active'
                    ? SmokeDetected.SMOKE_DETECTED
                    : SmokeDetected.SMOKE_NOT_DETECTED;
            },
        });
        this.registerCharacteristic({
            characteristicType: CarbonMonoxideDetected,
            serviceType: CarbonMonoxideSensor,
            getValue: (data) => {
                return data.co && data.co.alarmStatus === 'active'
                    ? CarbonMonoxideDetected.CO_LEVELS_ABNORMAL
                    : CarbonMonoxideDetected.CO_LEVELS_NORMAL;
            },
        });
        this.initSensorService(SmokeSensor);
        this.initSensorService(CarbonMonoxideSensor);
    }
}
exports.SmokeCoListener = SmokeCoListener;
