// homebridge-ring/lib/homebridge/temperature-sensor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemperatureSensor = void 0;
const base_device_accessory_1 = require("./base-device-accessory");
const hap_1 = require("./hap");
const operators_1 = require("rxjs/operators");
class TemperatureSensor extends base_device_accessory_1.BaseDeviceAccessory {
    constructor(device, accessory, logger, config) {
        super();
        this.device = device;
        this.accessory = accessory;
        this.logger = logger;
        this.config = config;
        const { Characteristic: { CurrentTemperature }, Service: { TemperatureSensor: TempSensorService }, } = hap_1.hap;
        this.registerObservableCharacteristic({
            characteristicType: CurrentTemperature,
            serviceType: TempSensorService,
            onValue: device.onData.pipe((0, operators_1.map)((data) => {
                return data.celsius;
            }), (0, operators_1.distinctUntilChanged)()),
        });
        this.initSensorService(TempSensorService);
    }
}
exports.TemperatureSensor = TemperatureSensor;
